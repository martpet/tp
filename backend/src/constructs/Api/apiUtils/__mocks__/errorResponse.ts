export const errorResponse = vi.fn().mockReturnValue('dummyerrorResponse');
