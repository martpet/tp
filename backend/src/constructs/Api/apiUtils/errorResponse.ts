import { getReasonPhrase, StatusCodes } from 'http-status-codes';
import { SetRequired } from 'type-fest';

type Options = WithExposedError | WithoutExposedError;

type Base = {
  description?: string;
  error?: unknown;
  statusCode?: StatusCodes;
};

type WithExposedError = SetRequired<Base, 'error'> & {
  expose: true;
};

type WithoutExposedError = Base & {
  expose?: never | false;
};

export const errorResponse = (
  traceId: string,
  {
    statusCode = StatusCodes.INTERNAL_SERVER_ERROR,
    description,
    error,
    expose,
  }: Options = {}
) => {
  const responseErrrorObject: Record<string, unknown> = {
    statusCode,
    message: getReasonPhrase(statusCode),
    trace_id: traceId,
  };

  if (description) {
    responseErrrorObject.description = description;
  }

  if (error instanceof Error) {
    console.error(`[${traceId}]`, error);

    if (expose) {
      responseErrrorObject.error = String(error);
    }
  }

  return {
    statusCode,
    body: JSON.stringify({ error: responseErrrorObject }),
  };
};
