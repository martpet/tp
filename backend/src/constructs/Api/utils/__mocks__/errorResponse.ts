export const errorResponse = vi.fn().mockReturnValue('dummyErrorResponse');
