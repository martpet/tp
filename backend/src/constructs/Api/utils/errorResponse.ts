import { APIGatewayProxyResult } from 'aws-lambda';
import { getReasonPhrase, StatusCodes } from 'http-status-codes';
import { SetRequired } from 'type-fest';

type ErrorResponseOptions = WithExposedError | WithoutExposedError;

type BaseOptions = {
  description?: string;
  error?: unknown;
  statusCode?: StatusCodes;
};

type WithExposedError = SetRequired<BaseOptions, 'error'> & {
  exposeError: true;
};

type WithoutExposedError = BaseOptions & {
  exposeError?: never | false;
};

export const errorResponse = (
  traceId: string,
  {
    statusCode = StatusCodes.INTERNAL_SERVER_ERROR,
    description,
    error,
    exposeError,
  }: ErrorResponseOptions = {}
): APIGatewayProxyResult => {
  const { envName } = globalLambdaProps;

  const errorObj: Record<string, unknown> = {
    statusCode,
    message: getReasonPhrase(statusCode),
    traceId,
  };

  if (description) {
    errorObj.description = description;
  }

  if (error) {
    console.error(`[${traceId}]`, error);

    if (exposeError ?? envName !== 'production') {
      errorObj.error = String(error);
    }
  }

  return {
    statusCode,
    body: JSON.stringify({ error: errorObj }),
  };
};
