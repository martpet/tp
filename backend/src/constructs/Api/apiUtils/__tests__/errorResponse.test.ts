import { StatusCodes } from 'http-status-codes';

import { errorResponse } from '../errorResponse';

const args: Parameters<typeof errorResponse> = ['dummyTraceId'];

describe('errorResponse', () => {
  it('returns a correct value', () => {
    expect(errorResponse(...args)).toMatchSnapshot();
  });

  describe('when "statusCode" is provided', () => {
    const argsClone = structuredClone(args);
    argsClone.push({ statusCode: StatusCodes.IM_A_TEAPOT });

    it('returns a correct value', () => {
      expect(errorResponse(...argsClone)).toMatchSnapshot();
    });
  });

  describe('when "description" is provided', () => {
    const argsClone = structuredClone(args);
    argsClone.push({ description: 'dummDescription' });

    it('returns a correct value', () => {
      expect(errorResponse(...argsClone)).toMatchSnapshot();
    });
  });

  describe('when "error" is provided', () => {
    const argsClone = structuredClone(args);
    argsClone.push({ error: new Error('dummyErrorMessage') });

    it('calls "console.error" with correct args', () => {
      errorResponse(...argsClone);
      expect(vi.mocked(console.error).mock.calls).toMatchSnapshot();
    });
  });

  describe('when "error" is provided and "expose" is true', () => {
    const argsClone = structuredClone(args);
    argsClone.push({
      error: new Error('dummyErrorMessage'),
      expose: true,
    });

    it('returns a correct value', () => {
      expect(errorResponse(...argsClone)).toMatchSnapshot();
    });
  });
});
