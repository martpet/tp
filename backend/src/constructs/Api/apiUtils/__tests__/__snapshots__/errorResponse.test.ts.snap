// Vitest Snapshot v1

exports[`errorResponse > returns a correct value 1`] = `
{
  "body": "{\\"error\\":{\\"statusCode\\":500,\\"message\\":\\"Internal Server Error\\",\\"trace_id\\":\\"dummyTraceId\\"}}",
  "statusCode": 500,
}
`;

exports[`errorResponse > when "description" is provided > returns a correct value 1`] = `
{
  "body": "{\\"error\\":{\\"statusCode\\":500,\\"message\\":\\"Internal Server Error\\",\\"trace_id\\":\\"dummyTraceId\\",\\"description\\":\\"dummDescription\\"}}",
  "statusCode": 500,
}
`;

exports[`errorResponse > when "error" is provided > calls "console.error" with correct args 1`] = `
[
  [
    "[dummyTraceId]",
    [Error: dummyErrorMessage],
  ],
]
`;

exports[`errorResponse > when "error" is provided and "expose" is true > returns a correct value 1`] = `
{
  "body": "{\\"error\\":{\\"statusCode\\":500,\\"message\\":\\"Internal Server Error\\",\\"trace_id\\":\\"dummyTraceId\\",\\"error\\":\\"Error: dummyErrorMessage\\"}}",
  "statusCode": 500,
}
`;

exports[`errorResponse > when "statusCode" is provided > returns a correct value 1`] = `
{
  "body": "{\\"error\\":{\\"statusCode\\":418,\\"message\\":\\"I'm a teapot\\",\\"trace_id\\":\\"dummyTraceId\\"}}",
  "statusCode": 418,
}
`;
